### PR TITLE
(fix): resolve data ordering to match axis for stacked violin plots

### LINKED ISSUE #3152: Incorrect column label and color assign when using 'rank_genes_groups_stacked_violin' with 'swap_axes=True'
### Please make sure these conditions are met

- [X] I have checked that this issue has not already been reported.
- [X] I have confirmed this bug exists on the latest version of scanpy.
- [X] (optional) I have confirmed this bug exists on the main branch of scanpy.

### What happened?

The column label and color is wrong after using 'swap_axes=True'
lyz1
lyz2

### Minimal code sample

```python
import scanpy as sc
adata = sc.datasets.pbmc68k_reduced()
sc.tl.rank_genes_groups(adata, 'bulk_labels', method='wilcoxon',tie_correct=True,pts=True,key_added='wilcoxon')
sc.pl.rank_genes_groups_stacked_violin(adata, n_genes=5, key="wilcoxon", groupby="bulk_labels")
sc.pl.rank_genes_groups_stacked_violin(adata, n_genes=5, key="wilcoxon", groupby="bulk_labels",swap_axes=True)
```

### Error output

_No response_

### Versions

<details>

```
-----
anndata     0.10.7
scanpy      1.10.2
-----
PIL                         10.3.0
anyio                       NA
arrow                       1.3.0
asttokens                   NA
astunparse                  1.6.3
attr                        23.2.0
attrs                       23.2.0
babel                       2.14.0
certifi                     2024.02.02
cffi                        1.16.0
charset_normalizer          3.3.2
cloudpickle                 3.0.0
comm                        0.2.2
cycler                      0.12.1
cython_runtime              NA
dask                        2024.5.2
dateutil                    2.9.0.post0
debugpy                     1.8.1
decorator                   5.1.1
defusedxml                  0.7.1
exceptiongroup              1.2.1
executing                   2.0.1
fastjsonschema              NA
fqdn                        NA
google                      NA
h5py                        3.11.0
idna                        3.7
igraph                      0.11.4
ipykernel                   6.29.4
isoduration                 NA
jedi                        0.19.1
jinja2                      3.1.3
joblib                      1.4.0
json5                       0.9.25
jsonpointer                 2.4
jsonschema                  4.21.1
jsonschema_specifications   NA
jupyter_events              0.10.0
jupyter_server              2.14.0
jupyterlab_server           2.27.0
kiwisolver                  1.4.5
legacy_api_wrap             NA
leidenalg                   0.10.2
llvmlite                    0.42.0
markupsafe                  2.1.5
matplotlib                  3.8.4
mpl_toolkits                NA
mpmath                      1.3.0
natsort                     8.4.0
nbformat                    5.10.4
numba                       0.59.1
numpy                       1.26.4
nvfuser                     NA
opt_einsum                  v3.3.0
overrides                   NA
packaging                   24.0
pandas                      1.5.3
parso                       0.8.4
platformdirs                4.2.0
plotly                      5.22.0
prometheus_client           NA
prompt_toolkit              3.0.43
psutil                      5.9.8
pure_eval                   0.2.2
pyarrow                     13.0.0
pydev_ipython               NA
pydevconsole                NA
pydevd                      2.9.5
pydevd_file_utils           NA
pydevd_plugins              NA
pydevd_tracing              NA
pygments                    2.17.2
pyparsing                   3.1.2
pythonjsonlogger            NA
pytz                        2024.1
referencing                 NA
requests                    2.31.0
rfc3339_validator           0.1.4
rfc3986_validator           0.1.1
rpds                        NA
scipy                       1.13.0
send2trash                  NA
session_info                1.0.0
six                         1.16.0
sklearn                     1.4.2
sniffio                     1.3.1
socks                       1.7.1
stack_data                  0.6.3
sympy                       1.12.1
tblib                       3.0.0
texttable                   1.7.0
threadpoolctl               3.4.0
tlz                         0.12.1
toolz                       0.12.1
torch                       2.0.0+cu117
tornado                     6.4
tqdm                        4.66.2
traitlets                   5.14.3
typing_extensions           NA
uri_template                NA
urllib3                     1.26.18
wcwidth                     0.2.13
webcolors                   1.13
websocket                   1.7.0
yaml                        6.0.1
zipp                        NA
zmq                         26.0.2
zoneinfo                    NA
-----
IPython             8.23.0
jupyter_client      8.6.1
jupyter_core        5.7.2
jupyterlab          4.1.6
-----
Python 3.10.14 (main, Apr 30 2024, 04:10:16) [GCC 13.2.1 20240417]
Linux-6.9.7-arch1-1-x86_64-with-glibc2.39
-----
Session information updated at 2024-07-12 14:59
```

</details>

### PR DESCRIPTION
**Pull request by ilan-gold**
- [x] Closes #3152
- [ ] Tests included or not required because:

- [ ] Release notes not necessary because:

### CODE REVIEW

**Comment by ilan-gold** [comment]
Will finish after lunch....wowza

**Comment by ilan-gold** [comment]
Not exactly sure how to test this - it's not that the axis is misordered, it's that we were not informing the violin plot of this ordering.  I am not sure if there is a way to access the underlying data of a plot...

**Comment by flying-sheep** [comment]
I have some limited experience with that plot order stuff.

There’s some compat code for old pandas versions added in #2816 which shows some spots where we mess with ordering.

Generally
- when user specifies an order, we use that
- if not, we rely on the DataFrame order for plotting, we don’t store this implicit order explicitly

**Comment by ilan-gold** [comment]
@flying-sheep I think the [code added in that PR](https://github.com/scverse/scanpy/pull/2816/files#diff-5d0e683154209be7830f09b5389551bf9700a4184d08e97c46c23e2e4beb54a0) is minimally relevant to what happened here.

> when user specifies an order, we use that

Right, so here the issue is that the category ordering is used for the labelling but we were not imposing it on the data itself when the violin plots render (separate from the axis labels, as the actual violin plots are added row-by-row).

> if not, we rely on the DataFrame order for plotting, we don’t store this implicit order explicitly

In some sense the above also applies.  If we want to add some sort of user-facing part of the API to allow for ordering, that is fine, but I think that should be separate as it would go into the next minor release and this is a fairly large bug.

I'm fine not testing this because I genuinely don't know how and I spent a few hours yesterday trying different things to no avail.

File: src/scanpy/plotting/_stacked_violin.py [hunk 1]
```diff
@@ -410,8 +410,30 @@ def _mainplot(self, ax: Axes):
...
+
+        # Ensure the categories axis is always ordered identically.
+        # If the axes are not swapped, the above _matrix.columns is used in the actual violin plot (i.e., unique names).
+        # If they are swapped, then use the same as the labels used below.
+        # Without this, `_make_rows_of_violinplots` does not know about the order of the categories in labels.
+        labels = _color_df.columns
+        if self.are_axes_swapped:
+            x_axis_order = labels
+        else:
+            x_axis_order = _matrix.columns
```
Comment (flying-sheep) [inline src/scanpy/plotting/_stacked_violin.py]:
```suggestion
        x_axis_order = labels if self.are_axes_swapped else _matrix.columns
```

**Review by flying-sheep — APPROVED** [review]
I trust you that this helps :shrug:

**Comment by ilan-gold** [comment]
> I trust you that this helps 🤷

I guess we can add a test that cements this behavior, but it's tough to test that swap=original-transpose for the underlying data.  I tried applying the ordering to what changed in the PR you referenced but that didn't help unfortunately.  I don't know the provenance of that change, so I think this is fair.  We are using the public API of seaborn in a way that fixes the underlying issue as one would expect.

**Pull request merged**